### PR TITLE
JSON and dictionary improvements

### PR DESCRIPTION
**Pull request by mpurland**

### CODE REVIEW

File: Swiftz/JSON.swift [hunk 1]
```diff
@@ -21,32 +21,30 @@ public enum JSONValue : CustomStringConvertible {
...
+			return s
+		case .JSONNull:
 			return NSNull()
 		}
 	}
 
 	// we know this is safe because of the NSJSONSerialization docs
-	private static func make(a : NSObject) -> JSONValue {
+	public static func make(a : AnyObject) -> JSONValue {
 		switch a {
```
Comment (CodaFi) [inline Swiftz/JSON.swift]: This should not be public.  It is not typesafe in all cases

File: Swiftz/JSON.swift [hunk 2]
```diff
@@ -459,7 +457,7 @@ public struct JDictionary<A, B : JSON where B.J == A> : JSON {
 
 /// MARK: Implementation Details
 
-private func resolveKeypath(lhs : Dictionary<String, JSONValue>, rhs : JSONKeypath) -> JSONValue? {
+public func resolveKeypath(lhs : Dictionary<String, JSONValue>, rhs : JSONKeypath) -> JSONValue? {
```
Comment (CodaFi) [inline Swiftz/JSON.swift]: This should not be public because it is SPI.

File: Swiftz/DictionaryExt.swift [hunk 3]
```diff
@@ -6,15 +6,28 @@
...
 //
 
-extension Dictionary {
-	/// Creates a Dictionary from a list of Key-Value pairs.
-	static func fromList(l : [(Key, Value)]) -> Dictionary<Key, Value> {
-		var d = Dictionary<Key, Value>(minimumCapacity: l.count)
-		l.forEach { (k, v) in
+/// Creates a dictionary from the elements array given a combiner function to return key and value for the dictionary.
+public func arrayToDictionary<E, K, V>(elements : [E], combiner: E -> (K, V)?) -> [K : V] {
+	return elements.reduce([K: V](minimumCapacity: elements.count)) { (var d, e) in
```
Comment (CodaFi) [inline Swiftz/DictionaryExt.swift]: Why isn't this in the extension, or a private function?

File: Swiftz/JSON.swift (hunk 1 above)
Comment (mpurland) [inline Swiftz/JSON.swift]: I'll make it private it again, but will work on an implementation that is typesafe in all cases and allows returning a proper error. I'm thinking something similar to Aeson's eitherDecode using an Either<String?, JSONDecodingError> sort of type.

**Comment by mpurland** [comment]
Let me know if this works.

I updated it to just use l.mapKeyValues { $0 } because l.mapKeyValues(identity) doesn't seem to compile on the Travis CI server (Compiles on my machine using Xcode 7.1.1 though)

File: Swiftz/DictionaryExt.swift [hunk 4]
```diff
@@ -6,14 +6,24 @@
 //  Copyright (c) 2014 Maxwell Swadling. All rights reserved.
 //
 
+extension Array {
+	/// Creates a dictionary from the array given a transformer function that returns 
+	/// a (key, value) pair for the dictionary, if nil is returned then the value is
+	/// not added to the dictionary.
+	public func mapKeyValues<K, V>(f : Element -> (K, V)?) -> [K: V] {
```
Comment (CodaFi) [inline Swiftz/DictionaryExt.swift]: Is there a more vivid name for this combinator?  mapAssoc or mapAssociate, etc.?

Comment (mpurland) [inline Swiftz/DictionaryExt.swift]: Come to think of it a more fitting name might be zipMap or groupBy. A similar function is called groupBy in Scala, but to accomplish this similarly we could just take in a function f that returns a key k for the value v corresponding to that element. Pseudo code:

zip(map(f)).toDictionary

Thoughts?

Comment (CodaFi) [inline Swiftz/DictionaryExt.swift]: I've seen it called `label` somewhere.  I quite like that idea.

Comment (mpurland) [inline Swiftz/DictionaryExt.swift]: I haven't heard of label, but that could make sense in that you're applying a label (a key) to each value.

Comment (mpurland) [inline Swiftz/DictionaryExt.swift]: @CodaFi Just did another commit attempting to pull in the idea of label and pairs. I also found a cleaner way to create a Dictionary from a list.

**Pull request closed without merging**